File: SafeFlex/SafeFlex/SensorService.swift
import Foundation

struct SensorReading: Decodable, Sendable {
    let timestamp: Int
    let flex: Double
    let stability: Double
}

@Observable
class SensorService {
    var flex: Double = 0
    var stability: Double = 0
    var isConnected = false

    var flexPercent: Int {
        min(100, max(0, Int(flex / 1000.0 * 100)))
    }

    var stabilityPercent: Int {
        min(100, max(0, Int(stability / 5.0 * 100)))
    }

    private var webSocketTask: URLSessionWebSocketTask?
    private var connectionTask: Task<Void, Never>?
    private var serverAddress = ""
    private let decoder = JSONDecoder()

    func connect(to address: String) {
        disconnect()
        serverAddress = address
        let urlString = "ws://\(address)"
        guard let url = URL(string: urlString) else {
            print("[Sensor] Invalid URL: \(urlString)")
            return
        }
        print("[Sensor] Connecting to \(urlString)...")
        let session = URLSession(configuration: .default)
        webSocketTask = session.webSocketTask(with: url)
        webSocketTask?.resume()
        isConnected = true

        connectionTask = Task { await receiveMessages() }
    }

    func disconnect() {
        connectionTask?.cancel()
        webSocketTask?.cancel(with: .goingAway, reason: nil)
        webSocketTask = nil
        isConnected = false
    }

    private func receiveMessages() async {
        guard let task = webSocketTask else { return }

        do {
            while !Task.isCancelled {
                let message = try await task.receive()
                let data: Data? = switch message {
                case .string(let text): text.data(using: .utf8)
                case .data(let d): d
                @unknown default: nil
                }
                guard let data else { continue }
                guard let reading = try? decoder.decode(SensorReading.self, from: data) else {
                    print("[Sensor] Failed to decode: \(String(data: data, encoding: .utf8) ?? "?")")
                    continue
                }
                flex = reading.flex
                stability = reading.stability
            }
        } catch {
            print("[Sensor] Connection error: \(error)")
            guard !Task.isCancelled else { return }
            isConnected = false
            try? await Task.sleep(for: .seconds(2))
            guard !Task.isCancelled else { return }
            print("[Sensor] Reconnecting...")
            connect(to: serverAddress)
        }
    }
}

File: SafeFlexRN/src/services/SensorService.ts
export interface SensorReading {
  timestamp: number;
  flex: number;
  stability: number;
}

type Listener = (reading: SensorReading) => void;
type StatusListener = (connected: boolean) => void;

class SensorService {
  private ws: WebSocket | null = null;
  private listeners: Listener[] = [];
  private statusListeners: StatusListener[] = [];
  private serverAddress = '';
  private shouldReconnect = false;

  get isConnected(): boolean {
    return this.ws?.readyState === WebSocket.OPEN;
  }

  connect(address: string) {
    this.disconnect();
    this.serverAddress = address;
    this.shouldReconnect = true;
    const url = `ws://${address}`;
    console.log(`[Sensor] Connecting to ${url}...`);

    this.ws = new WebSocket(url);

    this.ws.onopen = () => {
      console.log('[Sensor] Connected');
      this.notifyStatus(true);
    };

    this.ws.onmessage = (event) => {
      try {
        const reading: SensorReading = JSON.parse(event.data);
        this.listeners.forEach((l) => l(reading));
      } catch (e) {
        console.log('[Sensor] Failed to decode:', event.data);
      }
    };

    this.ws.onerror = (e) => {
      console.log('[Sensor] Error:', e);
    };

    this.ws.onclose = () => {
      console.log('[Sensor] Disconnected');
      this.notifyStatus(false);
      if (this.shouldReconnect) {
        setTimeout(() => {
          if (this.shouldReconnect) {
            console.log('[Sensor] Reconnecting...');
            this.connect(this.serverAddress);
          }
        }, 2000);
      }
    };
  }

  disconnect() {
    this.shouldReconnect = false;
    this.ws?.close();
    this.ws = null;
    this.notifyStatus(false);
  }

  onReading(listener: Listener) {
    this.listeners.push(listener);
    return () => {
      this.listeners = this.listeners.filter((l) => l !== listener);
    };
  }

  onStatusChange(listener: StatusListener) {
    this.statusListeners.push(listener);
    return () => {
      this.statusListeners = this.statusListeners.filter((l) => l !== listener);
    };
  }

  private notifyStatus(connected: boolean) {
    this.statusListeners.forEach((l) => l(connected));
  }
}

export const sensorService = new SensorService();
